nixos/modules/installer/tools/tools.nix

```diff
@@ -10,119 +10,163 @@ let
+          name = lib.mkOption {
+            type = lib.types.str;
+            description = lib.mdDoc ''
+              the name of the package
+            '';
+          };
+
+          enable = lib.mkEnableOption (lib.mdDoc "the package as part of the installer tools.");
+
+          package = lib.mkOption {
+            type = lib.types.package;
+            description = lib.mdDoc ''
+              the derivation providing the package
+            '';
+          };
+        };
+        config = {
+          name = lib.mkDefault name;
+        };
+      }));
     };
   };
 
-  options.system.disableInstallerTools = mkOption {
-    internal = true;
-    type = types.bool;
-    default = false;
-    description = lib.mdDoc ''
-      Disable nixos-rebuild, nixos-generate-config, nixos-installer
-      and other NixOS tools. This is useful to shrink embedded,
-      read-only systems which are not expected to be rebuild or
-      reconfigure themselves. Use at your own risk!
-    '';
-  };
-
-  config = lib.mkIf (config.nix.enable && !config.system.disableInstallerTools) {
+  config =
+  let
+    installerTools = lib.filterAttrs (_: tool: tool.enable) config.system.installerTools;
+  in
+  lib.mkIf (config.nix.enable && !config.system.disableInstallerTools) {
+
+    system.installerTools = {
+      nixos-build-vms = {
+        enable = true;
+        package = makeProg {
+          name = "nixos-build-vms";
+          src = ./nixos-build-vms/nixos-build-vms.sh;
+          inherit (pkgs) runtimeShell;
+        };
+      };
+
+      nixos-enter = {
+        enable = true;
+        package = makeProg {
+          name = "nixos-enter";
+          src = ./nixos-enter.sh;
+          inherit (pkgs) runtimeShell;
+          path = makeBinPath [
+            pkgs.util-linuxMinimal
+          ];
+        };
+      };
+
+      nixos-generate-config = {
+        enable = true;
+        package = makeProg {
+          name = "nixos-generate-config";
+          src = ./nixos-generate-config.pl;
+          perl = "${pkgs.perl.withPackages (p: [ p.FileSlurp ])}/bin/perl";
+          hostPlatformSystem = pkgs.stdenv.hostPlatform.system;
+          detectvirt = "${config.systemd.package}/bin/systemd-detect-virt";
+          btrfs = "${pkgs.btrfs-progs}/bin/btrfs";
+          inherit (config.system.nixos-generate-config) configuration desktopConfiguration;
+          xserverEnabled = config.services.xserver.enable;
+        };
+      };
+
+      nixos-install = {
+        enable = true;
+        package = makeProg {
+          name = "nixos-install";
+          src = ./nixos-install.sh;
+          inherit (pkgs) runtimeShell;
+          nix = config.nix.package.out;
+          path = makeBinPath [
+            pkgs.jq
+            config.system.installerTools.nixos-enter.package
+            pkgs.util-linuxMinimal
+          ];
+        };
+      };
+
+      nixos-option = {
+        enable = lib.versionOlder (lib.getVersion config.nix.package) "2.4pre";
```

Comment:
Good that I saw that, removed it in #237442 . Someone wants to review that?

Reply:
Hah, I looked for that PR because I remembered seeing that, but somehow it didn't show up in the search results. I think it's easy to rebase either of them on top of the other, depending on which one gets merged first.

I can look at the other suggestions you made in a bit.

Reply:
Feel free to merge this PR first, rebasing my change is not hard.